thirdparty/github.go

```diff
@@ -215,44 +260,34 @@ func githubShouldRetry(caller string, config retryConfig) utility.HTTPRetryFunct
 	}
 }
 
-// getGithubClient returns an *http.Client that will retry according to supplied retryConfig.
-// Also creates a span tracking the lifespan of the client.
-// Defer the returned function to release the client to the pool and close the span.
-func getGithubClient(ctx context.Context, token, caller string, config retryConfig, attributes []attribute.KeyValue) (context.Context, *http.Client, func()) {
+func getGithubClient(token, caller string, config retryConfig) *github.Client {
 	grip.Info(message.Fields{
 		"ticket":  GithubInvestigation,
 		"message": "called getGithubClient",
 		"caller":  caller,
 	})
 
-	ctx, span := tracer.Start(ctx, caller, trace.WithAttributes(append(attributes, attribute.String(githubEndpointAttribute, caller))...))
-
-	client := utility.GetHTTPClient()
-	originalTransport := client.Transport
-	client.Transport = otelhttp.NewTransport(client.Transport)
-
-	return ctx,
-		utility.SetupOauth2CustomHTTPRetryableClient(
-			token,
-			githubShouldRetry(caller, config),
-			utility.RetryHTTPDelay(utility.RetryOptions{
-				MaxAttempts: numGithubAttempts,
-				MinDelay:    githubRetryMinDelay,
-			}),
-			client,
-		), func() {
-			client.Transport = originalTransport
-			utility.PutHTTPClient(client)
-
-			span.End()
-		}
+	client := utility.SetupOauth2CustomHTTPRetryableClient(
+		token,
+		githubShouldRetry(caller, config),
+		utility.RetryHTTPDelay(utility.RetryOptions{
+			MaxAttempts: numGithubAttempts,
+			MinDelay:    githubRetryMinDelay,
+		}),
+		utility.DefaultHttpClient(cachingTransport),
+	)
+
+	return github.NewClient(client)
 }
 
-// GetInstallationToken creates an installation token using Github app auth.
-func GetInstallationToken(ctx context.Context, owner, repo string, opts *github.InstallationTokenOptions) (string, error) {
+// getInstallationToken creates an installation token using Github app auth.
+// If creating a token fails it will return the legacyToken.
+func getInstallationToken(ctx context.Context, legacyToken, owner, repo string, opts *github.InstallationTokenOptions) string {
+	trace.SpanFromContext(ctx).SetAttributes(attribute.Bool(githubAppTokenAttribute, false))
```

Comment:
can we do this after getting config?

Reply:
Oh or could you explain why we do false here and true later?

Reply:
The idea here was to add an attribute to the span for the GitHub call so we could count the ratio of the which token got used by route. `SetAttributes` overwrites any previous attributes with the same key so this set it to false originally and only true if the app token is used.